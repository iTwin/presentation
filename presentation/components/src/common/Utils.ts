/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Core */

import * as React from "react";
import { I18NNamespace } from "@bentley/imodeljs-i18n";
import { Presentation } from "@bentley/presentation-frontend";

/**
 * An interface of something that has a priority.
 * @internal
 */
export interface IPrioritized {
  priority: number;
}

/**
 * A sorting algorithm for `Array.sort` that sorts items by priority.
 * Higher priority items appear first in the list.
 *
 * @internal
 */
export const prioritySortFunction = (a: IPrioritized, b: IPrioritized): number => {
  if (a.priority > b.priority)
    return -1;
  if (a.priority < b.priority)
    return 1;
  return 0;
};

let localizationNamespace: I18NNamespace | undefined;
/**
 * Translate a string with the specified id from `PresentationComponents`
 * localization namespace. The `stringId` should not contain namespace - it's
 * prepended automatically.
 *
 * @internal
 */
export const translate = async (stringId: string): Promise<string> => {
  const localizationNamespaceName = "PresentationComponents";
  if (!localizationNamespace) {
    localizationNamespace = Presentation.i18n.registerNamespace(localizationNamespaceName);
  }
  await localizationNamespace.readFinished;
  stringId = `${localizationNamespaceName}:${stringId}`;
  return Presentation.i18n.translate(stringId);
};

/**
 * Creates a display name for the supplied component
 * @internal
 */
export const getDisplayName = <P>(component: React.ComponentType<P>): string => {
  if (component.displayName)
    return component.displayName;
  if (component.name)
    return component.name;
  return "Component";
};
