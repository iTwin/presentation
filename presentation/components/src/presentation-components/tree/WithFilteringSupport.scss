/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

.filteredTree {

  position: relative;
  width: 100%;
  height: 100%;

  .filteredTreeOverlay {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 9999;
    background-color: rgba(255,255,255,.8);
  }

}
