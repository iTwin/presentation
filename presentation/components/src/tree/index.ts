/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tree */

export { default as PresentationTreeDataProvider } from "./DataProvider";
export { default as IPresentationTreeDataProvider } from "./IPresentationTreeDataProvider";
export { default as withUnifiedSelection, SelectionTarget } from "./WithUnifiedSelection";
export { default as withFilteringSupport } from "./WithFilteringSupport";
