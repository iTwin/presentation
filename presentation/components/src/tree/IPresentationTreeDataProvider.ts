/*---------------------------------------------------------------------------------------------
* Copyright (c) 2019 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tree */

import { ITreeDataProvider, TreeNodeItem } from "@bentley/ui-components";
import { IPresentationDataProvider } from "../common/IPresentationDataProvider";
import { NodeKey, NodePathElement } from "@bentley/presentation-common";

/**
 * Presentation tree data provider.
 * @public
 */
export interface IPresentationTreeDataProvider extends ITreeDataProvider, IPresentationDataProvider {
  /**
   * Returns a [[NodeKey]] from given [[TreeNodeItem]].
   */
  getNodeKey(node: TreeNodeItem): NodeKey;

  /**
   * Returns filtered node paths.
   */
  getFilteredNodePaths(filter: string): Promise<NodePathElement[]>;

  /**
   * Loads the hierarchy so on-demand requests and filtering works quicker
   * @alpha
   */
  loadHierarchy?(): Promise<void>;
}
