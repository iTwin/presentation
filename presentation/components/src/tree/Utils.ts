/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Tree
 */

import { StyleHelper } from "../common/StyleHelper";
import { CheckBoxState } from "@bentley/ui-core";
import { Node, PageOptions as PresentationPageOptions, LabelDefinition } from "@bentley/presentation-common";
import { DelayLoadedTreeNodeItem, PageOptions as UiPageOptions, ItemStyle, ItemColorOverrides } from "@bentley/ui-components";
import { createLabelRecord } from "../common/Utils";

/** @internal */
export const PRESENTATION_TREE_NODE_KEY = "__presentation-components/key";

/** @internal */
export const createTreeNodeItems = (nodes: ReadonlyArray<Readonly<Node>>, parentId?: string): DelayLoadedTreeNodeItem[] => {
  const list = new Array<DelayLoadedTreeNodeItem>();
  for (const node of nodes)
    list.push(createTreeNodeItem(node, parentId));
  return list;
};

/** @internal */
export const createTreeNodeItem = (node: Readonly<Node>, parentId?: string): DelayLoadedTreeNodeItem => {
  const item: DelayLoadedTreeNodeItem = {
    id: [...node.key.pathFromRoot].reverse().join("/"),
    label: node.label,
    labelDefinition: createLabelRecord(node.labelDefinition ? node.labelDefinition : LabelDefinition.fromLabelString(node.label), "node_label"),
  };
  (item as any)[PRESENTATION_TREE_NODE_KEY] = node.key;

  const style: ItemStyle = {};
  const colorOverrides: ItemColorOverrides = {};
  if (parentId)
    item.parentId = parentId;
  if (node.description)
    item.description = node.description;
  if (node.hasChildren)
    item.hasChildren = true;
  if (node.isExpanded)
    item.autoExpand = true;
  if (node.imageId)
    item.icon = node.imageId;
  if (StyleHelper.isBold(node))
    style.isBold = true;
  if (StyleHelper.isItalic(node))
    style.isItalic = true;
  const foreColor = StyleHelper.getForeColor(node);
  if (foreColor)
    colorOverrides.color = foreColor;
  const backColor = StyleHelper.getBackColor(node);
  if (backColor)
    colorOverrides.backgroundColor = backColor;
  if (node.isCheckboxVisible) {
    item.isCheckboxVisible = true;
    if (node.isChecked)
      item.checkBoxState = CheckBoxState.On;
    if (!node.isCheckboxEnabled)
      item.isCheckboxDisabled = true;
  }
  if (Object.keys(colorOverrides).length > 0)
    style.colorOverrides = colorOverrides;
  if (Object.keys(style).length > 0)
    item.style = style;
  if (node.extendedData)
    item.extendedData = node.extendedData;

  return item;
};

/** @internal */
export const pageOptionsUiToPresentation = (pageOptions?: UiPageOptions): PresentationPageOptions | undefined => {
  if (pageOptions)
    return { ...pageOptions };
  return undefined;
};
