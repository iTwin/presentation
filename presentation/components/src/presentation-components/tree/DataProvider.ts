/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module Tree
 */

import memoize from "micro-memoize";
import { IDisposable, Logger } from "@bentley/bentleyjs-core";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { HierarchyRequestOptions, NodeKey, NodePathElement, Ruleset } from "@bentley/presentation-common";
import { Presentation } from "@bentley/presentation-frontend";
import { DelayLoadedTreeNodeItem, PageOptions, TreeNodeItem } from "@bentley/ui-components";
import { RulesetRegistrationHelper } from "../common/RulesetRegistrationHelper";
import { IPresentationTreeDataProvider } from "./IPresentationTreeDataProvider";
import { createTreeNodeItems, pageOptionsUiToPresentation, PRESENTATION_TREE_NODE_KEY } from "./Utils";

/**
 * Properties for creating a `PresentationTreeDataProvider` instance.
 * @public
 */
export interface PresentationTreeDataProviderProps {
  /** IModel to pull data from. */
  imodel: IModelConnection;
  /** Id of the ruleset to use when requesting content or a ruleset itself. */
  ruleset: string | Ruleset;
  /**
   * Paging size for obtaining nodes.
   *
   * Presentation data providers, when used with paging, have ability to save one backend request for size / count. That
   * can only be achieved when `pagingSize` property is set on the data provider and it's value matches size which is used when
   * requesting nodes. To help developers notice this problem, data provider emits a warning similar to this:
   * ```
   * PresentationTreeDataProvider.pagingSize doesn't match pageOptions in PresentationTreeDataProvider.getNodes call. Make sure you set PresentationTreeDataProvider.pagingSize to avoid excessive backend requests.
   * ```
   * To fix the issue, developers should make sure the page size used for requesting data is also set for the data provider:
   * ```TS
   * const pagingSize = 10;
   * const provider = new TreeDataProvider({imodel, ruleset, pagingSize});
   * // only one backend request is made for the two following requests:
   * provider.getNodesCount();
   * provider.getNodes({ start: 0, size: pagingSize });
   * ```
   */
  pagingSize?: number;
}

/**
 * Presentation Rules-driven tree data provider.
 * @public
 */
export class PresentationTreeDataProvider implements IPresentationTreeDataProvider, IDisposable {
  private _imodel: IModelConnection;
  private _rulesetRegistration: RulesetRegistrationHelper;
  private _pagingSize?: number;

  /** Constructor. */
  public constructor(props: PresentationTreeDataProviderProps) {
    this._rulesetRegistration = new RulesetRegistrationHelper(props.ruleset);
    this._imodel = props.imodel;
    this._pagingSize = props.pagingSize;
  }

  /** Destructor. Must be called to clean up.  */
  public dispose() {
    this._rulesetRegistration.dispose();
  }

  /** Id of the ruleset used by this data provider */
  public get rulesetId(): string { return this._rulesetRegistration.rulesetId; }

  /** [[IModelConnection]] used by this data provider */
  public get imodel(): IModelConnection { return this._imodel; }

  /**
   * Paging options for obtaining nodes.
   * @see `PresentationTreeDataProviderProps.pagingSize`
   */
  public get pagingSize(): number | undefined { return this._pagingSize; }
  public set pagingSize(value: number | undefined) { this._pagingSize = value; }

  /** Called to get extended options for node requests */
  private createRequestOptions(): HierarchyRequestOptions<IModelConnection> {
    return {
      imodel: this._imodel,
      rulesetOrId: this._rulesetRegistration.rulesetId,
    };
  }

  /**
   * Returns a [[NodeKey]] from given [[TreeNodeItem]].
   * **Warning:** the `node` must be created by this data provider.
   */
  public getNodeKey(node: TreeNodeItem): NodeKey {
    return (node as any)[PRESENTATION_TREE_NODE_KEY];
  }

  /**
   * Returns nodes
   * @param parentNode The parent node to return children for.
   * @param pageOptions Information about the requested page of data.
   */
  public async getNodes(parentNode?: TreeNodeItem, pageOptions?: PageOptions): Promise<DelayLoadedTreeNodeItem[]> {
    if (undefined !== pageOptions && pageOptions.size !== this.pagingSize) {
      const msg = `PresentationTreeDataProvider.pagingSize doesn't match pageOptions in PresentationTreeDataProvider.getNodes call.
        Make sure you set PresentationTreeDataProvider.pagingSize to avoid excessive backend requests.`;
      Logger.logWarning("Presentation.Components", msg);
    }

    return (await this._getNodesAndCount(parentNode, pageOptions)).nodes;
  }

  /**
   * Returns the total number of nodes
   * @param parentNode The parent node to return children count for.
   */
  public async getNodesCount(parentNode?: TreeNodeItem): Promise<number> {
    const pageOptions = undefined !== this.pagingSize ? { start: 0, size: this.pagingSize } : undefined;
    return (await this._getNodesAndCount(parentNode, pageOptions)).count;
  }

  private _getNodesAndCount = memoize(async (parentNode?: TreeNodeItem, pageOptions?: PageOptions): Promise<{ nodes: TreeNodeItem[], count: number }> => {
    const requestCount = undefined !== pageOptions && 0 === pageOptions.start && undefined !== pageOptions.size;
    const parentKey = parentNode ? this.getNodeKey(parentNode) : undefined;

    if (!requestCount) {
      const allNodes = await Presentation.presentation.getNodes({ ...this.createRequestOptions(), paging: pageOptionsUiToPresentation(pageOptions) }, parentKey);
      return { nodes: parentNode ? createTreeNodeItems(allNodes, parentNode.id) : createTreeNodeItems(allNodes), count: allNodes.length };
    }

    const nodesResponse = await Presentation.presentation.getNodesAndCount({ ...this.createRequestOptions(), paging: pageOptionsUiToPresentation(pageOptions) }, parentKey);
    return { nodes: parentNode ? createTreeNodeItems(nodesResponse.nodes, parentNode.id) : createTreeNodeItems(nodesResponse.nodes), count: nodesResponse.count };
  }, { isMatchingKey: MemoizationHelpers.areNodesRequestsEqual as any });

  /**
   * Returns filtered node paths.
   * @param filter Filter.
   */
  public getFilteredNodePaths = async (filter: string): Promise<NodePathElement[]> => {
    return Presentation.presentation.getFilteredNodePaths(this.createRequestOptions(), filter);
  }

  /**
   * Loads the hierarchy so on-demand requests and filtering works quicker
   * @alpha Hierarchy loading performance needs to be improved before this becomes publicly available.
   */
  public async loadHierarchy() {
    return Presentation.presentation.loadHierarchy(this.createRequestOptions());
  }

}

class MemoizationHelpers {
  public static areNodesRequestsEqual(lhsArgs: [TreeNodeItem?, PageOptions?], rhsArgs: [TreeNodeItem?, PageOptions?]): boolean {
    if (lhsArgs[0]?.id !== rhsArgs[0]?.id)
      return false;
    if ((lhsArgs[1]?.start ?? 0) !== (rhsArgs[1]?.start ?? 0))
      return false;
    if ((lhsArgs[1]?.size ?? 0) !== (rhsArgs[1]?.size ?? 0))
      return false;
    return true;
  }
}
