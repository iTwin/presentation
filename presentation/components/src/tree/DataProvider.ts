/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
/** @module Tree */

import * as _ from "lodash";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { NodeKey, NodePathElement, HierarchyRequestOptions } from "@bentley/presentation-common";
import { Presentation } from "@bentley/presentation-frontend";
import { DelayLoadedTreeNodeItem, TreeNodeItem } from "@bentley/ui-components/lib/tree/TreeDataProvider";
import { PageOptions } from "@bentley/ui-components/lib/common/PageOptions";
import { createTreeNodeItems, pageOptionsUiToPresentation } from "./Utils";
import IPresentationTreeDataProvider from "./IPresentationTreeDataProvider";

/**
 * Presentation Rules-driven tree data provider.
 */
export default class PresentationTreeDataProvider implements IPresentationTreeDataProvider {
  private _rulesetId: string;
  private _connection: IModelConnection;

  /**
   * Constructor.
   * @param connection Connection to an imodel to pull data from.
   * @param rulesetId Id of the ruleset used by this data provider.
   */
  public constructor(connection: IModelConnection, rulesetId: string) {
    this._rulesetId = rulesetId;
    this._connection = connection;
  }

  /** Id of the ruleset used by this data provider */
  public get rulesetId(): string { return this._rulesetId; }

  /** [[IModelConnection]] used by this data provider */
  public get connection(): IModelConnection { return this._connection; }

  /** Called to get extended options for node requests */
  private createRequestOptions(): HierarchyRequestOptions<IModelConnection> {
    return {
      imodel: this._connection,
      rulesetId: this._rulesetId,
    };
  }

  /**
   * Returns a [[NodeKey]] from given [[TreeNodeItem]].
   * **Warning:** the `node` must be created by this data provider.
   */
  public getNodeKey(node: TreeNodeItem): NodeKey {
    return node.extendedData.key as NodeKey;
  }

  /**
   * Returns nodes
   * @param parentNode The parent node to return children for.
   * @param pageOptions Information about the requested page of data.
   */
  public getNodes = _.memoize(async (parentNode?: TreeNodeItem, pageOptions?: PageOptions): Promise<DelayLoadedTreeNodeItem[]> => {
    if (parentNode) {
      const parentKey = this.getNodeKey(parentNode);
      const childNodes = await Presentation.presentation.getChildren({ ...this.createRequestOptions(), paging: pageOptionsUiToPresentation(pageOptions) }, parentKey);
      return createTreeNodeItems(childNodes, parentNode.id);
    }
    const rootNodes = await Presentation.presentation.getRootNodes({ ...this.createRequestOptions(), paging: pageOptionsUiToPresentation(pageOptions) });
    return createTreeNodeItems(rootNodes);
  }, MemoizationHelpers.getNodesKeyResolver);

  /**
   * Returns the total number of nodes
   * @param parentNode The parent node to return children count for.
   */
  public getNodesCount = _.memoize(async (parentNode?: TreeNodeItem): Promise<number> => {
    if (parentNode) {
      const parentKey = this.getNodeKey(parentNode);
      return await Presentation.presentation.getChildrenCount(this.createRequestOptions(), parentKey);
    } else {
      return await Presentation.presentation.getRootNodesCount(this.createRequestOptions());
    }
  }, MemoizationHelpers.getNodesCountKeyResolver);

  /**
   * Returns filtered node paths.
   * @param filter Filter.
   */
  public getFilteredNodePaths = async (filter: string): Promise<NodePathElement[]> => {
    return Presentation.presentation.getFilteredNodePaths(this.createRequestOptions(), filter);
  }
}

class MemoizationHelpers {
  public static createKeyForPageOptions(pageOptions?: PageOptions) {
    if (!pageOptions)
      return "0/0";
    return `${(pageOptions.start) ? pageOptions.start : 0}/${(pageOptions.size) ? pageOptions.size : 0}`;
  }
  public static createKeyForTreeNodeItem(item?: TreeNodeItem) { return item ? item.id : ""; }
  public static getNodesKeyResolver(parent?: TreeNodeItem, pageOptions?: PageOptions) {
    return `${MemoizationHelpers.createKeyForTreeNodeItem(parent)}/${MemoizationHelpers.createKeyForPageOptions(pageOptions)}`;
  }
  public static getNodesCountKeyResolver(parent?: TreeNodeItem) { return MemoizationHelpers.createKeyForTreeNodeItem(parent); }
}
