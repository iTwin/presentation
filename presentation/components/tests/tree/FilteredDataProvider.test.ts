/*---------------------------------------------------------------------------------------------
* Copyright (c) 2018 Bentley Systems, Incorporated. All rights reserved.
* Licensed under the MIT License. See LICENSE.md in the project root for license terms.
*--------------------------------------------------------------------------------------------*/
import { expect } from "chai";
import * as moq from "@bentley/presentation-common/tests/_helpers/Mocks";
import * as faker from "faker";
import { createRandomTreeNodeItem } from "../_helpers/UiComponents";
import { createRandomECInstanceNodeKey, createRandomNodePathElement } from "@bentley/presentation-common/tests/_helpers/random";
import { NodePathElement } from "@bentley/presentation-common";
import { PageOptions } from "@bentley/ui-components";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import FilteredPresentationTreeDataProvider from "../../lib/tree/FilteredDataProvider";
import IPresentationTreeDataProvider from "../../lib/tree/IPresentationTreeDataProvider";
import { createTreeNodeItem } from "../../lib/tree/Utils";

describe("FilteredTreeDataProvider", () => {

    /*
    A-1
      A-1-1
    A-2
      A-2-1
      A-2-2
        A-2-2-1
    */
    const nodePaths: NodePathElement[] = [];

    nodePaths[0] = createRandomNodePathElement();
    nodePaths[0].node.label = "A-1";

    nodePaths[1] = createRandomNodePathElement();
    nodePaths[1].node.label = "A-2";

    nodePaths[0].children = [];
    nodePaths[0].children[0] = createRandomNodePathElement();
    nodePaths[0].children[0].node.label = "A-1-1";

    nodePaths[1].children = [];
    nodePaths[1].children[0] = createRandomNodePathElement();
    nodePaths[1].children[0].node.label = "A-2-1";

    nodePaths[1].children[1] = createRandomNodePathElement();
    nodePaths[1].children[1].node.label = "A-2-2";

    nodePaths[1].children[1].children = [];
    nodePaths[1].children[1].children[0] = createRandomNodePathElement();
    nodePaths[1].children[1].children[0].node.label = "A-2-2-1";

    let provider: FilteredPresentationTreeDataProvider;
    let filter: string;
    const parentProviderMock = moq.Mock.ofType<IPresentationTreeDataProvider>();
    const imodelMock = moq.Mock.ofType<IModelConnection>();
    const pageOptions: PageOptions = { size: 0, start: 0 };

    beforeEach(() => {
        parentProviderMock.reset();
        filter = faker.random.word();
        provider = new FilteredPresentationTreeDataProvider(parentProviderMock.object, filter, nodePaths);
    });

    describe("filter", () => {

        it("returns filter with which it was initialized", () => {
            expect(provider.filter).to.be.equal(filter);
        });

    });

    describe("rulesetId", () => {

        it("returns rulesetId of the parent data provider", () => {
            const expectedRulesetId = faker.random.word();
            parentProviderMock.setup((x) => x.rulesetId)
                .returns(() => expectedRulesetId)
                .verifiable();
            expect(provider.rulesetId).to.eq(expectedRulesetId);
            parentProviderMock.verifyAll();
        });

    });

    describe("connection", () => {

        it("returns connection of the parent data provider", () => {
            parentProviderMock.setup((x) => x.connection)
                .returns(() => imodelMock.object)
                .verifiable();
            expect(provider.connection).to.eq(imodelMock.object);
            parentProviderMock.verifyAll();
        });

    });

    describe("getRootNodes", () => {

        it("return root nodes", async () => {
            const result = await provider.getRootNodes(pageOptions);
            expect(result).to.matchSnapshot();
        });

    });

    describe("getRootNodesCount", () => {

        it("returns root nodes count", async () => {
            const result = await provider.getRootNodesCount();
            expect(result).to.equal(nodePaths.length);
        });

    });

    describe("getChildNodes", () => {

        it("returns child nodes", async () => {
            const parentNode = createTreeNodeItem(nodePaths[1].node);

            const result = await provider.getChildNodes(parentNode, pageOptions);
            expect(result).to.matchSnapshot();
        });

    });

    describe("getChildNodesCount", () => {

        it("returns child node count", async () => {
            const parentNode = createTreeNodeItem(nodePaths[1].node);

            const result = await provider.getChildNodesCount(parentNode);
            expect(result).to.equal(nodePaths[1].children.length);
        });

    });

    describe("getFilteredNodePaths", () => {

        it("calls parent data provider", async () => {
            parentProviderMock.setup((x) => x.getFilteredNodePaths(filter))
                .returns(async () => nodePaths)
                .verifiable();

            const result = await provider.getFilteredNodePaths(filter);
            expect(result).to.equal(nodePaths);
            parentProviderMock.verifyAll();
        });

    });

    describe("getNodeKey", () => {

        it("returns node key", () => {
            const key = createRandomECInstanceNodeKey();
            const treeNode = createRandomTreeNodeItem(key);

            const result = provider.getNodeKey(treeNode);
            expect(result).to.deep.equal(key);
        });

    });

    describe("getAllNodeIds", () => {

        it("returns all ids for all nodes", () => {
            const allNodeIds = provider.getAllNodeIds();
            expect(allNodeIds).to.matchSnapshot();
        });

    });

    describe("", () => {
        const constantFilter = "test";
        const filteredNodePaths: NodePathElement[] = [];

        filteredNodePaths[0] = createRandomNodePathElement();
        filteredNodePaths[0].node.label = "A-1";
        filteredNodePaths[0].filteringData = { occurances: 0, childrenOccurances: 1 };

        filteredNodePaths[0].children = [];
        filteredNodePaths[0].children[0] = createRandomNodePathElement();
        filteredNodePaths[0].children[0].node.label = "A-1-1 test";
        filteredNodePaths[0].children[0].filteringData = { occurances: 1, childrenOccurances: 0 };

        filteredNodePaths[1] = createRandomNodePathElement();
        filteredNodePaths[1].node.label = "A-2 test";
        filteredNodePaths[1].filteringData = { occurances: 1, childrenOccurances: 0 };

        filteredNodePaths[1].children = [];
        filteredNodePaths[1].children[0] = createRandomNodePathElement();
        filteredNodePaths[1].children[0].node.label = "A-2-1";
        filteredNodePaths[1].children[0].filteringData = { occurances: 0, childrenOccurances: 0 };

        filteredNodePaths[1].children[1] = createRandomNodePathElement();
        filteredNodePaths[1].children[1].node.label = "A-2-2";
        filteredNodePaths[1].children[1].filteringData = { occurances: 0, childrenOccurances: 0 };

        describe("countFilteringResults", () => {
            it("all occurances get counted", () => {
                expect(provider.countFilteringResults(filteredNodePaths)).to.be.eq(2);
            });

            it("Doesn't count if nodePaths don't have filtering data", () => {
                expect(provider.countFilteringResults(nodePaths)).to.be.eq(0);
            });
        });

        describe("getActiveResultNode", () => {
            it("returns correct active node", () => {
                provider = new FilteredPresentationTreeDataProvider(parentProviderMock.object, constantFilter, filteredNodePaths);
                const activeResultNode = provider.getActiveResultNode(2);

                expect(activeResultNode).to.not.be.undefined;
                expect(activeResultNode!.id).to.be.eq(createTreeNodeItem(filteredNodePaths[1].node).id);
                expect(activeResultNode!.index).to.be.eq(0);
            });

            it("returns undefined when index is 0 or lower", () => {
                provider = new FilteredPresentationTreeDataProvider(parentProviderMock.object, constantFilter, filteredNodePaths);
                const activeResultNode = provider.getActiveResultNode(0);

                expect(activeResultNode).to.be.undefined;
            });
        });
    });
});
