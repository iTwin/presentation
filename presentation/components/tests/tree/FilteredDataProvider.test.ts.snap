// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilteredTreeDataProvider getAllNodeIds returns all ids for all nodes 1`] = `
Array [
  "be3b713c-8c3e-40b8-9206-010831d4838f/ec940c92-5980-4eb5-bec9-704edb8f1272",
  "e67862f3-ccbf-4c51-b8ed-ed1d0420ea19/6fbe024f-2316-4265-a6e8-d65a837e308a",
  "142b23f9-5f0d-4b34-b39f-de2c0a107d9c/dcb68625-021e-40e1-839b-08a0214f5943",
  "0a6e90df-c64f-4192-9396-a54b9bc1d0cc/c9427f95-0f93-463c-953e-d9823c990006",
  "729087ec-e569-4f02-9212-b8007ffd5072/9451bb8c-be2e-4a06-bb9c-f6fe83e0ae64",
  "fb7680c8-0a28-4cf0-9ee8-272ec6682ae5/e14e4128-01ae-43e4-8790-9992f94b3186",
]
`;

exports[`FilteredTreeDataProvider getChildNodes returns child nodes 1`] = `
Array [
  Object {
    "checkBoxState": 1,
    "description": "Sint libero laudantium voluptatem magni et voluptas quis.",
    "displayCheckBox": true,
    "extendedData": Object {
      "key": Object {
        "instanceKey": Object {
          "className": "Minnesota",
          "id": "0xcfbe0000007bf8",
        },
        "pathFromRoot": Array [
          "dcb68625-021e-40e1-839b-08a0214f5943",
          "142b23f9-5f0d-4b34-b39f-de2c0a107d9c",
        ],
        "type": "ECInstanceNode",
      },
    },
    "hasChildren": false,
    "id": "142b23f9-5f0d-4b34-b39f-de2c0a107d9c/dcb68625-021e-40e1-839b-08a0214f5943",
    "isCheckBoxEnabled": true,
    "label": "A-2-1",
    "labelBackColor": 1531058431,
    "labelBold": false,
    "labelForeColor": -1590574337,
    "labelItalic": false,
    "parentId": "fb7680c8-0a28-4cf0-9ee8-272ec6682ae5/e14e4128-01ae-43e4-8790-9992f94b3186",
  },
  Object {
    "checkBoxState": 1,
    "description": "",
    "displayCheckBox": false,
    "extendedData": Object {
      "key": Object {
        "instanceKey": Object {
          "className": "Incredible",
          "id": "0x332b000000c483",
        },
        "pathFromRoot": Array [
          "9451bb8c-be2e-4a06-bb9c-f6fe83e0ae64",
          "729087ec-e569-4f02-9212-b8007ffd5072",
        ],
        "type": "ECInstanceNode",
      },
    },
    "hasChildren": true,
    "id": "729087ec-e569-4f02-9212-b8007ffd5072/9451bb8c-be2e-4a06-bb9c-f6fe83e0ae64",
    "isCheckBoxEnabled": false,
    "label": "A-2-2",
    "labelBackColor": undefined,
    "labelBold": false,
    "labelForeColor": undefined,
    "labelItalic": false,
    "parentId": "fb7680c8-0a28-4cf0-9ee8-272ec6682ae5/e14e4128-01ae-43e4-8790-9992f94b3186",
  },
]
`;

exports[`FilteredTreeDataProvider getRootNodes return root nodes 1`] = `
Array [
  Object {
    "checkBoxState": 1,
    "description": "Repellendus fugiat velit sit aperiam placeat fuga doloribus.",
    "displayCheckBox": false,
    "extendedData": Object {
      "key": Object {
        "instanceKey": Object {
          "className": "bandwidth",
          "id": "0xa0da000001762c",
        },
        "pathFromRoot": Array [
          "6fbe024f-2316-4265-a6e8-d65a837e308a",
          "e67862f3-ccbf-4c51-b8ed-ed1d0420ea19",
        ],
        "type": "ECInstanceNode",
      },
    },
    "hasChildren": true,
    "id": "e67862f3-ccbf-4c51-b8ed-ed1d0420ea19/6fbe024f-2316-4265-a6e8-d65a837e308a",
    "isCheckBoxEnabled": false,
    "label": "A-1",
    "labelBackColor": undefined,
    "labelBold": false,
    "labelForeColor": undefined,
    "labelItalic": false,
    "parentId": undefined,
  },
  Object {
    "checkBoxState": 1,
    "description": "Voluptas enim ex eveniet facere.",
    "displayCheckBox": false,
    "extendedData": Object {
      "key": Object {
        "instanceKey": Object {
          "className": "front-end",
          "id": "0xf9270000012549",
        },
        "pathFromRoot": Array [
          "e14e4128-01ae-43e4-8790-9992f94b3186",
          "fb7680c8-0a28-4cf0-9ee8-272ec6682ae5",
        ],
        "type": "ECInstanceNode",
      },
    },
    "hasChildren": true,
    "id": "fb7680c8-0a28-4cf0-9ee8-272ec6682ae5/e14e4128-01ae-43e4-8790-9992f94b3186",
    "isCheckBoxEnabled": false,
    "label": "A-2",
    "labelBackColor": 499774463,
    "labelBold": false,
    "labelForeColor": -284526081,
    "labelItalic": false,
    "parentId": undefined,
  },
]
`;
